libbpf-tools/funclatency.c

```diff
@@ -0,0 +1,370 @@
+// SPDX-License-Identifier: (LGPL-2.1 OR BSD-2-Clause)
+/* Copyright (c) 2021 Google LLC.
+ *
+ * Based on funclatency from BCC by Brendan Gregg and others
+ * 2021-02-26   Barret Rhoden   Created this.
+ *
+ * TODO:
+ * - support uprobes on libraries without -p PID. (parse ld.so.cache)
+ * - support uprobes on all processes on all cpus (libbpf change)
+ */
+#include <argp.h>
+#include <errno.h>
+#include <signal.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <time.h>
+#include <unistd.h>
+
+#include <bpf/libbpf.h>
+#include <bpf/bpf.h>
+#include "funclatency.h"
+#include "funclatency.skel.h"
+#include "trace_helpers.h"
+#include "map_helpers.h"
+#include "uprobe_helpers.h"
+
+/* Adapted from linux/err.h */
+#define MAX_ERRNO 4095
+#define ERR_PTR(err) ((void *)((intptr_t)(err)))
+#define PTR_ERR(ptr) ((intptr_t)(ptr))
+#define IS_ERR(x) ((unsigned long)(void *)(x) >= (unsigned long)-MAX_ERRNO)
+#define IS_ERR_OR_NULL(ptr) (!(ptr) || IS_ERR(ptr))
+
+#define handle_error(msg) \
+	do { perror(msg); exit(EXIT_FAILURE); } while (0)
+#define warn(...) fprintf(stderr, __VA_ARGS__)
+
+static struct prog_env {
+	int units;
+	pid_t pid;
+	unsigned int duration;
+	unsigned int interval;
+	unsigned int iterations;
+	bool timestamp;
+	char *funcname;
+} env = {
+	.interval = 99999999,
+	.iterations = 99999999,
+};
+
+const char *argp_program_version = "funclatency 0.1";
+const char *argp_program_bug_address = "<bpf@vger.kernel.org>";
+static const char args_doc[] = "FUNCTION";
+static const char program_doc[] =
+"Time functions and print latency as a histogram\n"
+"\n"
+"Usage: funclatency [-h] [-m|-u] [-p PID] [-d DURATION] [ -i INTERVAL ]\n"
+"                   [-T] FUNCTION\n"
+"       Choices for FUNCTION: FUNCTION         (kprobe)\n"
+"                             LIBRARY:FUNCTION (uprobe a library in -p PID)\n"
+"                             :FUNCTION        (uprobe the binary of -p PID)\n"
+// TODO: Currently broken; see above
+//"                             PROGRAM:FUNCTION (uprobe the binary PROGRAM)\n"
+"\v"
+"Examples:\n"
+"  ./funclatency do_sys_open         # time the do_sys_open() kernel function\n"
+"  ./funclatency -m do_nanosleep     # time do_nanosleep(), in milliseconds\n"
+"  ./funclatency -u vfs_read         # time vfs_read(), in microseconds\n"
+"  ./funclatency -p 181 vfs_read     # time process 181 only\n"
+"  ./funclatency -p 181 c:read       # time the read() C library function\n"
+"  ./funclatency -p 181 :foo         # time foo() from pid 181's userspace\n"
+"  ./funclatency -i 2 -d 10 vfs_read # output every 2 seconds, for 10s\n"
+"  ./funclatency -mTi 5 vfs_read     # output every 5 seconds, with timestamps\n"
+;
+
+static const struct argp_option opts[] = {
+	{ "milliseconds", 'm', NULL, 0, "Output in milliseconds"},
+	{ "microseconds", 'u', NULL, 0, "Output in microseconds"},
+	{0, 0, 0, 0, ""},
```

Comment:
what are these empty slots for?

Reply:
They help with sorting the arg options by breaking the options into 'groups'.  Argp sorts them alphabetically within these groups.  The empty slot will make it so that -m and -u are next to each other.